lib/commons/dom/get-element-stack.js

```diff
@@ -0,0 +1,488 @@
+		);
+	}
+	if (isStackingContext(vNode)) {
+		stackingOrder.push(0);
+	}
+
+	return stackingOrder;
+}
+
+/**
+ * Determine if a node is a scroll region.
+ * @param {VirtualNode}
+ * @return {Boolean}
+ */
+function isScrollRegion(vNode) {
+	const style = vNode.computedStyle;
+
+	return (
+		style.getPropertyValue('overflow-x') === 'auto' ||
+		style.getPropertyValue('overflow-x') === 'scroll' ||
+		style.getPropertyValue('overflow-y') === 'auto' ||
+		style.getPropertyValue('overflow-y') === 'scroll'
+	);
+}
+
+/**
+ * Return the parent node that is a scroll region.
+ * @param {VirtualNode}
+ * @return {VirtualNode|null}
+ */
+function findScrollRegionParent(vNode) {
+	let scrollRegionParent = null;
+	let vNodeParent = vNode.parent;
+	let checkedNodes = [vNode];
+
+	while (vNodeParent) {
+		if (vNodeParent._scrollRegionParent) {
+			scrollRegionParent = vNodeParent._scrollRegionParent;
+			break;
+		}
+
+		if (isScrollRegion(vNodeParent)) {
+			scrollRegionParent = vNodeParent;
+			break;
+		}
+
+		checkedNodes.push(vNodeParent);
+		vNodeParent = vNodeParent.parent;
+	}
+
+	// cache result of parent scroll region so we don't have to look up the entire
+	// tree again for a child node
+	checkedNodes.forEach(
+		vNode => (vNode._scrollRegionParent = scrollRegionParent)
+	);
+	return scrollRegionParent;
+}
+
+/**
+ * Get the DOMRect x or y value. IE11 (and Phantom) does not support x/y
+ * on DOMRect.
+ * @param {DOMRect}
+ * @param {String} pos 'x' or 'y'
+ * @return {Number}
+ */
+function getDomPosition(rect, pos) {
+	if (pos === 'x') {
+		return 'x' in rect ? rect.x : rect.left;
+	}
+
+	return 'y' in rect ? rect.y : rect.top;
+}
+
+/**
+ * Add a node to every cell of the grid it intersects with.
+ * @param {Grid}
+ * @param {VirtualNode}
+ */
+function addNodeToGrid(grid, vNode) {
+	// save a reference to where this element is in the grid so we
+	// can find it even if it's in a subgrid
+	vNode._grid = grid;
+
+	vNode.clientRects.forEach(rect => {
+		const startRow = Math.floor(getDomPosition(rect, 'y') / gridSize);
+		const startCol = Math.floor(getDomPosition(rect, 'x') / gridSize);
+
+		const endRow = Math.floor(
+			(getDomPosition(rect, 'y') + rect.height) / gridSize
+		);
+		const endCol = Math.floor(
+			(getDomPosition(rect, 'x') + rect.width) / gridSize
+		);
+
+		for (let row = startRow; row <= endRow; row++) {
+			grid.cells[row] = grid.cells[row] || [];
+
+			for (let col = startCol; col <= endCol; col++) {
+				grid.cells[row][col] = grid.cells[row][col] || [];
+
+				if (!grid.cells[row][col].includes(vNode)) {
+					grid.cells[row][col].push(vNode);
+				}
+			}
+		}
+	});
+}
+
+/**
+ * Setup the 2d grid and add every element to it.
+ */
+function createGrid() {
+	const rootGrid = {
+		container: null,
+		cells: []
+	};
+
+	axe.utils
+		.querySelectorAll(axe._tree[0], '*')
+		.filter(vNode => vNode.actualNode.parentElement !== document.head)
+		.forEach(vNode => {
+			if (vNode.actualNode.nodeType !== 1) {
```

Comment:
Minor nitpick, is `Node.ELEMENT_NODE` not an option here?

Reply:
It probably is, but we haven't used it before. If we did, we'd need to use it under `window` so we don't relay on `Node` as a global.

Reply:
Having magic numbers do make it less approachable, especially if I don't know what `nodeType` is coming from. Since this is sticking with existing conventions I'm not going to suggest a one-off change here, but it would be nice to eventually either alias node types to have more readable code.

Reply:
Node.ELEMENT_NODE should exist everywhere. If not, there are polyfills (like https://github.com/inexorabletash/polyfill/blob/HEAD/dom.js)